File: backend/api/src/main/java/cz/sendinel/api/entity/Sender.java
package cz.sendinel.api.entity;

import com.fasterxml.jackson.databind.JsonNode;
import cz.sendinel.shared.enums.SenderTypesEnum;
import jakarta.persistence.Column;
import jakarta.persistence.Entity;
import jakarta.persistence.FetchType;
import jakarta.persistence.GeneratedValue;
import jakarta.persistence.GenerationType;
import jakarta.persistence.Id;
import jakarta.persistence.JoinColumn;
import jakarta.persistence.ManyToOne;
import jakarta.persistence.OneToMany;
import jakarta.persistence.Table;
import lombok.Getter;
import lombok.Setter;
import org.hibernate.annotations.ColumnDefault;
import org.hibernate.annotations.JdbcTypeCode;
import org.hibernate.annotations.SQLDelete;
import org.hibernate.type.SqlTypes;

import java.time.Instant;
import java.util.LinkedHashSet;
import java.util.Set;
import java.util.UUID;

@Getter
@Setter
@Entity
@Table(name = "senders")
@SQLDelete(sql = "UPDATE senders SET deleted_on = CURRENT_TIMESTAMP WHERE id = ?")
public class Sender {
    @Id
    @GeneratedValue(strategy = GenerationType.UUID)
    @Column(name = "id", nullable = false)
    private UUID id;

    @Column(name = "name", length = 50)
    private String name;

    @Column(name = "type", length = 25)
    private SenderTypesEnum type;

    @Column(name = "description")
    private String description;

    @JdbcTypeCode(SqlTypes.JSON)
    @Column(name = "configuration", columnDefinition = "jsonb")
    private JsonNode configuration;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "created_by")
    private User createdBy;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "updated_by")
    private User updatedBy;

    @ColumnDefault("now()")
    @Column(name = "created_on", nullable = false)
    private Instant createdOn;

    @ColumnDefault("now()")
    @Column(name = "updated_on", nullable = false)
    private Instant updatedOn;

    @Column(name = "deleted_on")
    private Instant deletedOn;

    @OneToMany(mappedBy = "sentBy")
    private Set<Email> emails = new LinkedHashSet<>();

}

File: frontend/src/lib/types/dtos/sender.ts
import type { SenderConfigurationFieldType, SenderTypesEnum } from "../enums/sender-types-enum";
import type { UserBasicsResponse } from "./user";

export interface SenderBasicsResponse {
  id: string; // UUID
  name: string;
  description: string;
  type: SenderTypesEnum;
}

export interface SenderRequest {
  name: string;
  type: SenderTypesEnum;
  description: string;
  configuration: Record<string, SenderConfigurationFieldType>;
}

export interface SenderResponse {
  id: string; // UUID
  name: string;
  type: SenderTypesEnum;
  description: string;
  configuration: Record<string, SenderConfigurationFieldType>; // JsonNode
  createdBy: UserBasicsResponse;
  updatedBy: UserBasicsResponse;
  createdOn: string; // Instant
  updatedOn: string; // Instant
  deletedOn: string | null; // Instant
}
